### ISSUE TITLE
Feature request: Add noLeading option

### ISSUE THREAD

**Issue by ahderman** [niksy/throttle-debounce#34]
# Context

I currently use the `throttle()` function in an online form that autosaves the contents of `<input>` fields when they change. As a user types their text in the form, the throttled `save()` method is called once every five seconds.

In this scenario, the leading call to the throttled function simply saves the first character that was typed, which is not very useful. Nothing breaks because of it, but it's also a call that does not need to be made in my use-case.

# Proposal

Lodash has a boolean `leading` option that allows skipping the first invocation to the throttled function. I propose adding the same thing to this package.

**Comment by niksy** [comment]
PR welcome!

**Comment by gajus** [comment]
I thought I am going insane, trying every combination of options available to achieve this.

**Comment by niksy** [comment]
As I said, PR welcome!

**Comment by seiyab** [comment]
What the API should look like?
I think it is better to discuss it before PR.

## idea1
```js
function(delay, [noLeading], [noTrailing], callback, [debounceMode])

// example for noLeading
throttle(200, true, false, f)
```

For backward compatibility, we need to declare `noTrailing` when we use `noLeading`.
It is a bit confusing because single boolean between `delay` and `callback` will be assumed as `noTrailing` though `noLeading` is former option.

## idea2
```js
function(delay, [noTrailing], [noLeading], callback, [debounceMode])

// example for noLeading
throttle(200, false, true, f)
```

It avoids confusion that idea1 has.
However, it is still confusing because the order of `noTrailing` and `noLeading` doesn't follow time.

## idea3
```ts
function(delay, [noTrailing], callback, [debounceMode], [options])
type Options = {
  noLeading?: boolean;
  noTraling?: boolean;
  debounceMode?: boolean;
}

// example for noLeading
throttle(200, f, {noLeading: true})
```

It avoids both confusions above and also solves future problems similar to them.
Declaring optional setting with new `options` argument will produce readability.
But we have duplication of options for backward compatibility.

**Comment by niksy** [comment]
I’m leaning towards third option, but definitely with major version bump so we can remove duplicate arguments.

**Comment by seiyab** [comment]
Thank you for your response.

When I tried to implement, I encountered a problem.
As far as I know, all the callback executions on `noTrailing = true` is leading executions.
If we simply skip leading callback execution, `throttle` with `noLeading = true` and `noTraling = true` will never execute callback.
Is it OK?
Or do I misunderstand?
Here are available options I'm considering
- `throttle` with `noLeading = true` and `noTrailing = true` will never execute callback.
- `throttle` with `noLeading = true` and `noTrailing = true` will throw an error immediately.
- `noLeading = true` will skip leading **throttled-function** call rather than callback execution.

**Comment by niksy** [comment]
@seiyab yes, this makes sense, and should be documented to avoid confusion.

**Comment by seiyab** [comment]
I created a PR https://github.com/niksy/throttle-debounce/pull/49 .

**Comment by niksy** [comment]
This has been released in [v4.0.0](https://github.com/niksy/throttle-debounce/releases/tag/v4.0.0).

Thanks @seiyab for your work on this!
